pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,350 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package basicstationlns
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"time"
+
+	"github.com/gorilla/websocket"
+	echo "github.com/labstack/echo/v4"
+	"go.thethings.network/lorawan-stack/pkg/auth/rights"
+	"go.thethings.network/lorawan-stack/pkg/basicstation"
+	"go.thethings.network/lorawan-stack/pkg/errors"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io"
+	"go.thethings.network/lorawan-stack/pkg/gatewayserver/io/basicstationlns/messages"
+	"go.thethings.network/lorawan-stack/pkg/log"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/unique"
+	"go.thethings.network/lorawan-stack/pkg/web"
+	"google.golang.org/grpc/metadata"
+)
+
+var (
+	errEmptyGatewayEUI           = errors.Define("empty_gateway_eui", "empty gateway EUI")
+	errMessageTypeNotImplemented = errors.DefineUnimplemented("message_type_not_implemented", "message of type `{type}` is not implemented")
+)
+
+type srv struct {
+	ctx      context.Context
+	server   io.Server
+	upgrader *websocket.Upgrader
+	tokens   io.DownlinkTokens
+}
+
+func (*srv) Protocol() string { return "basicstation" }
+
+// New returns a new Basic Station frontend that can be registered in the web server.
+func New(ctx context.Context, server io.Server) web.Registerer {
+	ctx = log.NewContextWithField(ctx, "namespace", "gatewayserver/io/basicstation")
+	return &srv{
+		ctx:      ctx,
+		server:   server,
+		upgrader: &websocket.Upgrader{},
+	}
+}
+
+func (s *srv) RegisterRoutes(server *web.Server) {
+	group := server.Group(ttnpb.HTTPAPIPrefix + "/gs/io/basicstation")
+	group.GET("/discover", s.handleDiscover)
+	group.GET("/traffic/:id", s.handleTraffic)
+}
+
+func (s *srv) handleDiscover(c echo.Context) error {
+	ctx := c.Request().Context()
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "discover",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
```

Comment:
Feels a bit strange to log "failures" (as you call them) as debug level here and below. They should be warn. Also, I would write "could not" instead of "failed to", because this is something the server can't do because the client screwed up, not because the server failed.

Reply:
Let's not trigger log warnings from garbage requests from clients.

Regarding the message; it doesn't say the server failed, it says an action failed. The blame is in the cause.

We do not (consistently) differentiate "could not" vs "failed to" to indicate external vs internal. I also don't see that feasible to implement because it fully depends on the cause. And we should not be switching on cause codes to print the "right" log message.

Reply:
For the record: I'm still not quite happy with this, but ok